src/Microsoft.Unity.Analyzers/BaseMethodUsage.cs

```diff
@@ -0,0 +1,75 @@
+﻿/*--------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See LICENSE in the project root for license information.
+ *-------------------------------------------------------------------------------------------*/
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Microsoft.Unity.Analyzers
+{
+	[AttributeUsage(AttributeTargets.Method)]
+	public class MethodUsageAttribute : Attribute
+	{
+	}
+
+	public abstract class BaseMethodUsageAnalyzer<T> : DiagnosticAnalyzer where T : MethodUsageAttribute
+	{
+		private static ILookup<string, MethodInfo> _lookup;
+
+		public override void Initialize(AnalysisContext context)
+		{
+			context.EnableConcurrentExecution();
+			context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+			context.RegisterSyntaxNodeAction(AnalyzeInvocation, SyntaxKind.InvocationExpression);
+		}
+
+		protected virtual bool IsReportable(IMethodSymbol method)
+		{
+			if (_lookup == null)
+			{
+				_lookup = CollectMethods(GetType().Assembly)
```

Comment:
I think we're generalizing too early, and by doing so we end up with an implementation which is very open but also pretty inefficient. We're looking at every single method in the assembly for an attribute. Even if've re building the lookup once that's quite a bit of time and memory that's going to be used, and we'll be doing that for every type extending `MethodUsageAnalyzer`.

For the SetPixels implementation, we know in which types we need to find those attributes, and we even know the name of the methods. I think it would be interesting to have the implementor of a MethodUsageAnalyzer be able to reduce the amount of reflection done here. Maybe we could move the responsibility to provide the methods to lookup to the extender.